### ISSUE TITLE
feat(rust): fslc の Rust 移植設計 — Rust CLI + ブラウザ WASM(z3-solver npm ブリッジ)

### ISSUE THREAD

**Issue by rizumita** [ymm-oss/fsl#195]
## 背景と目的

fslc を Rust に移植し、(1) 単一バイナリのネイティブ CLI、(2) ブラウザ(Web Worker)で動く WASM ビルドの2ターゲットを得る。狙いは配布の単純化(venv/pip 不要)、`docs/intro/` に埋め込むブラウザプレイグラウンド、エージェントサンドボックスへの持ち込み。

本 Issue は現行実装の計測(Z3 API 使用範囲・CLI/JSON 契約・テスト装置・文法・モジュール規模)に基づく設計の確定版。計測は 2026-07-11 時点の working tree に基づく。

**計測サマリ**

- `src/fslc` 全体 ≈ 32.9k 行(`bmc.py` 5,908 / `cli.py` 2,295 / `dialects.py` 2,028 / `runtime.py` 1,703 / `refine.py` 1,445 / `html_report.py` 1,430 / `testgen.py` 1,328 / `model.py` 1,327 …)
- Z3 使用面は極小(付録A): ソートは Int/Bool/Array のみ。tactics / Optimize / native Exists / 未解釈関数 / timeout・seed 設定は不使用
- 文法は Lark **Earley**(左再帰あり・~150 規則)、AST はタプル、方言(requirements/business/governance/compose)は `build_spec` 前の純粋 AST 書き換え
- corpus snapshot(`tests/test_corpus_snapshot.py`)は判定射影のみ(トレース除外・sort_keys 正規化)→ **そのまま移植のパリティゲートに使える**
- `z3-solver` npm(公式 WASM)は `check()` のみ非同期(Asyncify)、項構築・モデル評価は同期。SharedArrayBuffer 必須(COOP/COEP)。公開中断 API 無し。Rust `z3`/`z3-sys` crate は vendored 静的リンク+バージョンピン可、wasm32 サポート無し → **ブラウザは npm ブリッジ一択**

## ゴール / 非ゴール

**ゴール**

1. fslc(Rust)が現行 JSON エンベロープ + exit code 契約を保存する(snapshot 射影レベルで完全一致)
2. dual evaluator(BMC ↔ Monitor)+ Z3 非依存 BFS オラクルという正当性装置ごと移植する
3. wasm32-unknown-unknown + wasm-bindgen で Web Worker からカーネルコマンドを実行し、ソルバは公式 `z3-solver` npm にブリッジする

**非ゴール(v1)**

- Z3 の Rust WASM への静的リンク(npm ブリッジで回避)
- wasm32-wasi 単一 wasm CLI(需要が出たら別 Issue)
- 言語仕様の変更・エンベロープへの新フィールド追加
- エンコーディング変更による性能改善(まず同型移植)

## 1. ワークスペース構成

同一リポジトリ内 `rust/` に cargo workspace を置く(corpus・snapshot・オラクルを両実装で共有するため)。

```
rust/
├─ fsl-syntax        字句+構文解析(手書き再帰下降+Pratt)、AST、位置情報
├─ fsl-core          build_spec・型→ソート・定数評価・方言展開(dialects/compose/governance)
│                    診断(faithfulness/trace_type)・エンベロープ整形・FileResolver
├─ fsl-runtime       Monitor / replay / BFS オラクル(★ solver 系クレート依存禁止)
├─ fsl-solver        ソルバ抽象 trait(§2)
│   ├─ fsl-solver-z3    native backend(z3 crate, vendored 静的リンク)
│   └─ fsl-solver-z3js  browser backend(wasm-bindgen → z3-solver npm)
├─ fsl-verifier      BMC / k-induction / refine / scenarios(trait 総称、async)
├─ fsl-tools         mutate / explain / testgen / typestate / html_report / acceptance / ledger / analyze
├─ fslc              CLI バイナリ(clap、エンベロープ出力、exit code)
└─ fsl-wasm          wasm-bindgen cdylib + Web Worker プロトコル
```

依存規則: `fsl-runtime → fsl-core → fsl-syntax` のみ。**fsl-runtime に solver 系クレートを依存させない**ことを CI で検査する。現行では「runtime は Z3 非依存」は規約でしかないが、Rust ではクレートグラフによる構造保証に格上げできる(dual evaluator + 独立オラクルという中核不変条件の型レベル化)。

## 2. ソルバ抽象 trait — 計測済み表面のみ

現行 `bmc.py`/`refine.py` が使う Z3 面は驚くほど小さい(付録A)。trait はこの表面「だけ」を持つ。疑似コード:

```rust
enum Sort { Bool, Int, Array(Box<Sort>, Box<Sort>) }
enum Value { Int(i64), Bool(bool) }
enum SatResult { Sat, Unsat, Unknown }

trait Smt {
    type Term: Clone;
    // 項構築 — 全て同期(z3-solver npm でも Z3_mk_* は同期)
    fn int_val(&self, v: i64) -> Term;
    fn bool_val(&self, b: bool) -> Term;
    fn constant(&self, name: &str, s: &Sort) -> Term;
    fn add/sub/mul/div/modulo(...) -> Term;   // div/mod は SMT-LIB(Euclidean)意味論
    fn eq/lt/le/gt/ge(...) -> Term;
    fn and_/or_(&self, ts: &[Term]) -> Term;
    fn not_/implies/ite(...) -> Term;
    fn select/store(...) -> Term;
    fn const_array(&self, s: &Sort, default: Term) -> Term;   // z3.K
    fn forall(&self, bound: &[Term], body: Term) -> Term;     // 型境界 lowering 2箇所のみ
    fn substitute(...) -> Term;                                // refine の抽象init置換 1箇所
    fn sort_of(&self, t: &Term) -> Sort;                       // 現行 sort().kind() 相当
    // ソルバ — check のみ async(z3-solver npm の Asyncify 境界)
    fn push(&mut self); fn pop(&mut self);
    fn assert(&mut self, t: Term);
    fn assert_and_track(&mut self, t: Term, label: &str);
    async fn check(&mut self) -> SatResult;
    async fn check_assuming(&mut self, lits: &[Term]) -> SatResult; // requires-coverage 診断 1箇所
    fn unsat_core(&self) -> Vec<String>;
    fn model_eval(&self, t: &Term) -> Value;   // model_completion=true 意味論(同期)
    fn version(&self) -> String;               // faithfulness / verify_cache キー用
}
```

設計判断:

- ソルバパラメータは `unsat_core` 有効化のみ(現行と同一)。timeout/seed/logic は現行も未使用 → 決定性が高く、backend 間の verdict 一致に有利
- `z3.simplify` は定数畳み込み 1 箇所のみ → Rust 側ネイティブの定数畳み込みで置換し trait から外す
- 値は `i64` + checked arithmetic(オーバーフローは silent wrap でなくエラー)。FSL のドメインは有限小域なので実用上十分
- ユーザの `forall`/`exists` は現行どおり有限域 ground 展開(native quantifier 不使用)。native ForAll は set/map の型境界 lowering 2 箇所だけ

## 3. 非同期設計とキャンセル

- `fsl-verifier` は async。await 点は `check()` 境界のみ。native backend は同期 z3 呼び出しを即時完了 future で包む。CLI は `pollster::block_on` 程度で吸収
- BMC の「インクリメンタル1ソルバ + depth 毎に遷移を恒久 add + プロパティ毎 `push / add(Not(prop)) / check / pop`」構造は同型移植(付録A)
- ブラウザのキャンセル: `z3-solver` npm に公開中断 API が無い(実測)→ 検証は専用 Web Worker 内で実行し、**キャンセル = `worker.terminate()` + 再生成**(z3 再 init)。check 完了後のスレッド残留には `em.PThread.terminateAllThreads()` を適用
- depth ループ各周回で progress イベントを post(プレイグラウンドの UX 要件)

## 4. パーサ戦略

- Lark Earley(左再帰・~150 規則)は LALR 系ジェネレータへ機械移植不可 → **手書き再帰下降 + Pratt**。現行文法の式は既に precedence cascade(`expr → quant → implies → or → and → not → is → cmp → sum → product → unary → postfix → atom`)なので素直に写せる
- 位置情報(line/column)を全ノードに保持。パースエラーは `kind:"parse"` + `loc` を厳密一致、`expected`/`message` 文言は best-effort(snapshot が固定するのは kind までであることを確認済み)
- AST は Rust enum(現行タプル AST と 1:1 対応)。方言展開(dialects/compose/governance)は AST→AST 純関数として同型移植
- 検収: Python 側にタプル AST の JSON ダンプ機能を足し、corpus 全ファイルで AST 構造差分を取る(Phase 0)
- **FileResolver trait**: `parse_src(base_dir)` 経由の兄弟ファイル読み(`use ... from`, `implements ... from`)を抽象化。native = fs、wasm = リクエスト同梱の in-memory map

## 5. 値ドメインと dual evaluator

Monitor(fsl-runtime)と BMC(fsl-verifier)は現行同様、物理状態レイアウトを共有する:

- 物理キー命名 `__present` / `__value` / `__data` / `__len` / `__<field>` を厳密踏襲
- enum は int 序数、Set/Relation/Map は有限域の密表現、Seq は data+len、整数除算・剰余は Euclidean(両評価器一致が corpus 一致の前提)
- 論理(表示)状態への変換(`phys_to_logical` 相当)も同型移植し、トレース JSON スキーマ `{step, state, action:{name,params}, changes}` を保存

移植する検査装置:

- BFS オラクル(`tests/oracle.py` 相当)の Rust 化 + Python 版もアービタとして残す
- expression-agreement(到達可能な具象状態を pin して `sym != conc` の unsat 証明)の Rust 化

## 6. JSON / exit-code 契約の保存

- エンベロープ: `"fsl":"1.0"` 先頭・挿入順・indent=2・非 ASCII 生値(`ensure_ascii=False` 相当)・faithfulness のネスト dict への再帰 setdefault・`trace_type` のトップレベル末尾 setdefault。serde_json(preserve_order)+ **CLI と Worker が同一の emit 実装を共有**
- exit code 表(0/1/2/3)、`mutate` の常時 exit 0、`chain` が `_envelope` を通らない(faithfulness/trace_type 無し)+ テーブルの stderr 出力、raw stdout バイパス群 — をすべて仕様として列挙し再現(付録B)

## 7. パリティ検証ハーネス(移行ゲート)

3 層:

1. **snapshot 射影一致** — `tests/test_corpus_snapshot.py` の射影を Rust 出力に適用し、既存 golden と厳密一致(トレース除外・sort_keys 正規化は現行仕様のまま流用)
2. **全エンベロープ構造 diff** — 許容差分 allowlist(トレース内容・エラー文言)以外はフィールド単位一致
3. **トレース相互 replay** — Rust の counterexample を Python Monitor で replay(逆も)し、実 violation になることを検証(Z3 witness の非一意性を吸収)

加えて、Rust オラクル vs Rust BMC の verdict 一致(`assert_verdict_agrees` 相当)と、Python オラクル vs Rust BMC のクロス。CI はこのハーネスを Phase ゲートとして常時実行する。

## 8. WASM / ブラウザ設計

- ターゲット: wasm32-unknown-unknown + wasm-bindgen。`fsl-wasm` は Worker 内で動く
- Worker プロトコル: `{id, cmd, source, files?, options}` → `{id, envelope}` + `{id, progress}`。`chain`(`[impl]` が subprocess)と `-o` 系ファイル出力はブラウザ対象外(コマンド allowlist)
- `z3-solver` npm: `check` のみ async、項構築・model eval は同期(実測)→ §2 の trait 設計と整合
- SharedArrayBuffer 必須 → COOP/COEP。**GitHub Pages はレスポンスヘッダ設定不可 → `coi-serviceworker` を `docs/intro/` に同梱**
- ペイロード: `z3-solver` は unpacked ≈33MB(wasm 本体が大半)。プレイグラウンドは遅延ロード + Service Worker キャッシュで受容
- 同じ npm ブリッジは Node(≥16)でも動く → npm 配布で Node 環境カバーは将来 Issue

## 9. バージョン固定と faithfulness

- Z3 は native(z3-sys の vendored / gh-release + `Z3_SYS_Z3_VERSION`)と browser(`z3-solver` npm)を**同一リリースにピン**する。提案: 4.16.0 系で統一し、移行期間中は pip の z3-solver も同版に揃える
- verify_cache のキーに solver version + backend 識別子を含める。エンベロープへの新フィールド追加は v1 ではしない(契約凍結)

## 10. フェーズ計画

| Phase | 内容 | ゲート |
|---|---|---|
| 0 スパイク | (a) wasm-bindgen ↔ z3-solver npm 往復(項構築→check→model eval を Worker 内で) (b) Rust パーサ × corpus AST 差分 (c) JS ブリッジの項構築スループット計測 | 技術リスクの実証 |
| 1 カーネル | fsl-syntax / fsl-core(kernel spec のみ)/ fsl-runtime / fsl-verifier(BMC verify)/ check・verify・replay・scenarios / エンベロープ | `specs/` の snapshot 射影一致 + オラクル一致 |
| 2 全言語面 | induction / refine / dialects(requirements・business・governance)/ compose / sweep / chain | corpus 全体の snapshot 一致 + 相互 replay |
| 3 周辺 | mutate / explain / testgen / typestate / html_report / acceptance / ledger / analyze / db・ai・domain フロントエンド | サブコマンド別パリティ + Rust 側回帰テスト |
| 4 WASM | fsl-wasm / Worker プロトコル / docs/intro プレイグラウンド(coi-serviceworker) | ブラウザで corpus サンプルの verify が CLI と同一 verdict |

規模の目安(Python LOC 換算): Phase 1 ≈ 8k、Phase 2 ≈ +7k、Phase 3 ≈ +12k(db/ai/domain 含む・要精査)。合計 ≈33k 行の同型移植であり、フェーズ単位で独立に着手・検収できる構成にしてある。

## 11. 移行期間の Python 版の扱い

- Python 版は**リファレンス実装 + アービタ**として維持。移植期間中のカーネル意味論変更は原則凍結し、必要な変更は「両実装同時 + パリティゲート通過」を条件とする
- Rust が Phase 3 ゲートを通過し数リリース分パリティを維持できた時点で、Python を保守モードに移す(別 Issue で判断)

## 12. リスク / 未決事項

**リスク**

- Earley→再帰下降の受理言語差(潜在的曖昧性)→ Phase 0 の AST 差分 + corpus ベースの fuzz で潰す
- JS ブリッジの項構築オーバーヘッド(BMC は項を大量構築、1 呼び出し = 1 JS 呼び)→ Phase 0 で計測。過大なら assertion 単位の SMT-LIB テキスト一括投入を browser backend の代替実装として検討
- 移植中の本体進化(#174〜#193 のパイプライン)との二重実装コスト → §11 の凍結ポリシーで抑制
- z3-solver npm の中断 API 不在 → Worker 破棄で代替(確定)
- wasm ペイロード ≈33MB → 遅延ロード + キャッシュで受容

**未決事項**

1. v1 スコープ: db/ai/domain フロントエンド・testgen 6 ターゲット・html_report を初期移植に含めるか、当面 Python 残置とするか
2. Z3 ピンの具体バージョン(4.16.0 を提案)
3. 整数表現(i64 + checked arithmetic を提案)
4. `rust/` ワークスペースの CI 構成(既存 GitHub Actions への統合方法)

## 付録A: 計測済み Z3 API 表面(bmc.py 5,908 行 / refine.py 1,445 行)

- **ソート**: IntSort / BoolSort / ArraySort のみ。enum は int、option/struct/seq は物理変数へ field 展開。BitVec / Real / Datatype / SeqSort / 未解釈ソートは不使用
- **項**: 演算子オーバーロード(+ − × div mod・比較)、And/Or/Not/Implies/If、Select/Store/K。Distinct / Lambda / Exists / Function(未解釈関数)は不使用
- native ForAll は set/map の型境界 lowering **2 箇所のみ**。ユーザ量化子は有限域 ground 展開
- **ソルバ**: 素の `Solver()` のみ。パラメータは `unsat_core=True` のみ(timeout / seed / tactics / Optimize 不使用)
- **パターン**: (1) インクリメンタル 1 ソルバ + depth 毎に遷移を恒久 add + プロパティ毎 push/Not/check/pop(verify・induction とも)、(2) 使い捨てソルバ(liveness ランキング・診断・refine)
- `check(assumptions)` 1 箇所(requires-coverage 診断)、`assert_and_track` + `unsat_core` 2 箇所(到達性診断・requires 最小化)
- **モデル**: `model.eval(…, model_completion=True)` → `as_long` / `is_true` / `is_false` のみ
- その他: `substitute` 1 箇所(refine)、`simplify` 1 箇所(定数畳み込み)、`get_version_string`(cache キー)

## 付録B: 落としやすい契約の癖(実測)

- エンベロープは挿入順・indent=2・ensure_ascii=False。faithfulness は**ネストした dict にも再帰付与**(setdefault)。trace_type はトップレベルのみ末尾 setdefault
- `chain` は `_envelope` を通らない(fsl/result/layers… を手組み、faithfulness/trace_type 無し)+ chain テーブルは stderr へ
- `mutate` は結果に関わらず exit 0
- raw stdout バイパス: `version` / `typestate --ts` / `testgen`・`html`・`ledger`(`-o` 無し)/ `explain --readable` / `analyze` 非 json / `db import`(`imported` のみ。`imported_with_warnings` は JSON)/ `domain expand` / `domain testgen`
- Monitor の step 成功 verdict に `params` キーが無い(トレース生成側は params を呼び出し元から取る必要がある)
- 整数除算・剰余は Euclidean(`runtime` の `_euc_div` と SMT-LIB div の一致が dual evaluator の前提)
- snapshot は判定射影のみ(sort_keys / ensure_ascii=True の正準 JSON)。トレースは意図的に除外 → Z3 witness の非一意性はパリティゲートに影響しない

**Comment by rizumita** [comment]
PR #206 ではRust移植の設計文書だけをmainへ取り込みました。Rust実装は未着手です。Issueが自動closeされたのはPR本文の `Closes #195` によるもので、実装作業を継続するためreopenしました。次は設計書のPhase 0（WASM↔z3-solverスパイク、Rust parser/corpus AST差分、JS bridge throughput計測）から進めます。
